scene/2d/parallax_2d.cpp

```diff
@@ -0,0 +1,275 @@
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "parallax_2d.h"
+#include "core/config/project_settings.h"
+
+void Parallax2D::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+#ifdef TOOLS_ENABLED
+			if (Engine::get_singleton()->is_editor_hint()) {
+				set_notify_local_transform(true);
+			}
+
+#endif // TOOLS_ENABLED
+			group_name = "__cameras_" + itos(get_viewport_rid().get_id());
+			add_to_group(group_name);
+			_update_repeat();
+			_update_scroll();
+		} break;
+
+		case NOTIFICATION_INTERNAL_PROCESS: {
+			if (Engine::get_singleton()->is_editor_hint() || (!repeat_size.x && !repeat_size.y) || (!autoscroll.x && !autoscroll.y)) {
+				return;
+			}
+
+			autoscroll_offset += autoscroll * get_process_delta_time();
+			autoscroll_offset = autoscroll_offset.posmodv(repeat_size);
+
+			_update_scroll();
+		} break;
+
+#ifdef TOOLS_ENABLED
+		case NOTIFICATION_LOCAL_TRANSFORM_CHANGED: {
+			if (scroll_dirty) {
+				scroll_dirty = false;
+			} else {
+				set_scroll_offset(get_position());
+			}
+		} break;
+
+#endif // TOOLS_ENABLED
+		case NOTIFICATION_EXIT_TREE: {
+			remove_from_group(group_name);
+		} break;
+	}
+}
+
+void Parallax2D::_camera_moved(const Transform2D &p_transform, const Point2 &p_screen_offset, const Point2 &p_adj_screen_pos) {
+	if (ignore_camera_scroll) {
+		return;
+	}
+
+	set_screen_offset(p_adj_screen_pos);
+}
+
+void Parallax2D::_update_scroll() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	Point2 scroll_ofs = screen_offset;
+	Size2 vps = get_viewport_rect().size;
+
+#ifdef TOOLS_ENABLED
+	if (Engine::get_singleton()->is_editor_hint()) {
+		vps = Size2(GLOBAL_GET("display/window/size/viewport_width"), GLOBAL_GET("display/window/size/viewport_height"));
+	}
+
+#endif // TOOLS ENABLED
```

Comment:
```suggestion
#endif // TOOLS ENABLED

```
Just maybe transpose these lines here.

Reply:
The reason I have the spacing where it is, is when the if define isn't used, it would either create an extra second new line, or an absence of one.

For example this one:
```cpp
	Size2 vps = get_viewport_rect().size;

#ifdef TOOLS_ENABLED
	if (Engine::get_singleton()->is_editor_hint()) {
		vps = Size2(GLOBAL_GET("display/window/size/viewport_width"), GLOBAL_GET("display/window/size/viewport_height"));
	}

#endif // TOOLS ENABLED
	if (limit_begin.x <= limit_end.x - vps.x) {
		scroll_ofs.x = CLAMP(scroll_ofs.x, limit_begin.x, limit_end.x - vps.x);
	}
```

Becomes this when the flag isn't enabled:
```cpp
	Size2 vps = get_viewport_rect().size;

	if (limit_begin.x <= limit_end.x - vps.x) {
		scroll_ofs.x = CLAMP(scroll_ofs.x, limit_begin.x, limit_end.x - vps.x);
	}
```

And with your change, it'd be this:

```cpp
	Size2 vps = get_viewport_rect().size;


	if (limit_begin.x <= limit_end.x - vps.x) {
		scroll_ofs.x = CLAMP(scroll_ofs.x, limit_begin.x, limit_end.x - vps.x);
	}
```

Reply:
I don't understand. Like, yeah, it would create a new line if `TOOLS_ENABLED` is not defined. But this code is compiled, so it doesn't matter in the end, does it?